### PR TITLE
Add getPublishersInfo method to Node

### PR DESCRIPTION
**Pull request by ivanpauno**
Depends on https://github.com/osrf/ros2_java/pull/25.
Depends on https://github.com/osrf/ros2_java/pull/30.

This PR adds:
* An `EndpointInfo` class, to describe the value return by `getPublishersInfo` and `getSubscriptionsInfo` (unimplemented).
* Implements `getPublishersInfo` with proper unit testing.

### CODE REVIEW

**Comment by ivanpauno** [comment]
Converting to draft until https://github.com/osrf/ros2_java/pull/30 and #25 are merged.

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+  jbyte * gid_content = env->GetByteArrayElements(jgid, nullptr);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  for (size_t i = 0; i < RMW_GID_STORAGE_SIZE; ++i) {
+    gid_content[i] = p->endpoint_gid[i];
+  }
+  env->ReleaseByteArrayElements(jgid, gid_content, 0);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  env->SetObjectField(self, endpoint_gid_fid, jgid);
+  jclass qos_clazz = env->FindClass("org/ros2/rcljava/qos/QoSProfile");
+  jmethodID qos_init_mid = env->GetMethodID(qos_clazz, "<init>", "()V");
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp]: Should we check for exceptions after each of these lines?

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  jobject jqos = env->NewObject(qos_clazz, qos_init_mid);
+  RCLJAVA_COMMON_CHECK_FOR_EXCEPTION(env);
+  jmethodID qos_from_rcl_mid = env->GetMethodID(qos_clazz, "nativeFromRCL", "(J)V");
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp:104]: Same here?

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp [hunk 3]
```diff
@@ -250,3 +252,59 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeCreateTimerHandle(
   jlong jtimer = reinterpret_cast<jlong>(timer);
   return jtimer;
 }
+
+JNIEXPORT void JNICALL
+Java_org_ros2_rcljava_node_NodeImpl_nativeGetPublishersInfo(
+  JNIEnv * env, jclass, jlong handle, jstring jtopic_name, jobject jpublishers_info)
+{
+  rcl_node_t * node = reinterpret_cast<rcl_node_t *>(handle);
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp:318]: Consider checking if null. I guess we'll throw an error on L281 if node is null, but it's probably not a bad idea to check anyways.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp [hunk 4]
```diff
@@ -250,3 +252,59 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeCreateTimerHandle(
...
+    for (size_t i = 0; i < publishers_info.size; i++) {
+      jobject item = env->NewObject(endpoint_info_clazz, endpoint_info_init_mid);
+      env->CallVoidMethod(item, endpoint_info_from_rcl_mid, &publishers_info.info_array[i]);
+      env->CallBooleanMethod(jpublishers_info, list_add_mid, item);
+      env->DeleteLocalRef(item);
+    }
+  }
+
+// TODO(ivanpauno): Add a dependency on rcpputils and use "scope exit" utility
+// instead of C style error handling (?).
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: I don't have a preference, I think rcpputils is a fairly light dependency if you want to introduce it. Either way, I think we can remove this TODO.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp [hunk 5]
```diff
@@ -250,3 +252,59 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeCreateTimerHandle(
...
+    jmethodID endpoint_info_init_mid = env->GetMethodID(endpoint_info_clazz, "<init>", "()V");
+    RCLJAVA_COMMON_CHECK_FOR_EXCEPTION_WITH_ERROR_STATEMENT(env, goto cleanup);
+    jmethodID endpoint_info_from_rcl_mid = env->GetMethodID(
+      endpoint_info_clazz, "nativeFromRCL", "(J)V");
+    RCLJAVA_COMMON_CHECK_FOR_EXCEPTION_WITH_ERROR_STATEMENT(env, goto cleanup);
+
+    for (size_t i = 0; i < publishers_info.size; i++) {
+      jobject item = env->NewObject(endpoint_info_clazz, endpoint_info_init_mid);
+      env->CallVoidMethod(item, endpoint_info_from_rcl_mid, &publishers_info.info_array[i]);
+      env->CallBooleanMethod(jpublishers_info, list_add_mid, item);
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Any of these statements can throw; we should probably check for that.

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp (hunk 1 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp]: Addressed in https://github.com/osrf/ros2_java/pull/26/commits/0024d68a62ed3809ff5cbaaccce06467f4e00f04.

File: rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp (hunk 2 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_graph_EndpointInfo.cpp:104]: Yeap, see https://github.com/osrf/ros2_java/pull/26/commits/0024d68a62ed3809ff5cbaaccce06467f4e00f04

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp (hunk 3 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp:318]: Done in https://github.com/osrf/ros2_java/pull/26/commits/0024d68a62ed3809ff5cbaaccce06467f4e00f04.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp (hunk 5 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Done in https://github.com/osrf/ros2_java/pull/26/commits/0024d68a62ed3809ff5cbaaccce06467f4e00f04.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp (hunk 4 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Sounds good, I will add the dependency tomorrow.

Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: See https://github.com/osrf/ros2_java/pull/26/commits/111fab9c4e363a060c55318e8b19d0163fd32ade.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp [hunk 6]
```diff
@@ -309,3 +310,66 @@ Java_org_ros2_rcljava_node_NodeImpl_nativeGetTopicNamesAndTypes(
...
 }
+
+JNIEXPORT void JNICALL
+Java_org_ros2_rcljava_node_NodeImpl_nativeGetPublishersInfo(
+  JNIEnv * env, jclass, jlong handle, jstring jtopic_name, jobject jpublishers_info)
+{
+  rcl_node_t * node = reinterpret_cast<rcl_node_t *>(handle);
+  if (!node) {
+    rcljava_throw_exception(
+      env, "java/lang/IllegalStateException", "passed node handle is NULL");
```
Comment (jacobperron) [inline rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: nit: I think this should be `IllegalArgumentException` (like other cases where we check for null handles).

**Review by jacobperron — APPROVED** [review]
LGTM. One minor comment.

File: rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp (hunk 6 above)
Comment (ivanpauno) [reply rcljava/src/main/cpp/org_ros2_rcljava_node_NodeImpl.cpp]: Done in https://github.com/osrf/ros2_java/pull/26/commits/73fba78b06322b03463109df756dbe2f16c778ec.

**Pull request merged**